ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplDumpTask.java

```diff
@@ -137,17 +137,69 @@ private void prepareReturnValues(List<String> values) throws SemanticException {
+   * locations.
+   * 2. External or ACID tables are being bootstrapped for the first time : so that we can dump
+   * those tables as a whole.
+   * @return
+   */
+  private boolean shouldDumpExternalTableLocation() {
+    return conf.getBoolVar(HiveConf.ConfVars.REPL_INCLUDE_EXTERNAL_TABLES)
+            && (!conf.getBoolVar(HiveConf.ConfVars.REPL_DUMP_METADATA_ONLY)
+            || conf.getBoolVar(HiveConf.ConfVars.REPL_BOOTSTRAP_EXTERNAL_TABLES));
+  }
+
+  private boolean shouldExamineTablesToDump() {
+    return shouldDumpExternalTableLocation() ||
+            conf.getBoolVar(HiveConf.ConfVars.REPL_BOOTSTRAP_ACID_TABLES);
+  }
+
+  private boolean shouldBootstrapDumpTable(Table table) {
+    if (conf.getBoolVar(HiveConf.ConfVars.REPL_BOOTSTRAP_EXTERNAL_TABLES) &&
+            TableType.EXTERNAL_TABLE.equals(table.getTableType())) {
+      return true;
+    }
+
+    if (conf.getBoolVar(HiveConf.ConfVars.REPL_BOOTSTRAP_ACID_TABLES) &&
+           AcidUtils.isTransactionalTable(table)) {
+      return true;
+    }
+
+    return false;
+  }
+
   private Long incrementalDump(Path dumpRoot, DumpMetaData dmd, Path cmRoot, Hive hiveDb) throws Exception {
     Long lastReplId;// get list of events matching dbPattern & tblPattern
     // go through each event, and dump out each event to a event-level dump dir inside dumproot
+    ValidTxnList validTxnList = null;
+    long waitUntilTime = 0;
+    long bootDumpBeginReplId = -1;
+
+    // If we are bootstrapping ACID tables, we need to perform steps similar to a regular
+    // bootstrap (See bootstrapDump() for more details. Only difference here is instead of
+    // waiting for the concurrent transactions to finish, we start dumping the incremental events
+    // and wait only for the remaining time if any.
+    if (conf.getBoolVar(HiveConf.ConfVars.REPL_BOOTSTRAP_ACID_TABLES)) {
+      bootDumpBeginReplId = queryState.getConf().getLong(ReplUtils.LAST_REPL_ID_KEY, -1L);
+      assert (bootDumpBeginReplId >= 0);
+      LOG.info("Dump for bootstrapping ACID tables during an incremental dump for db {} and table {}",
+              work.dbNameOrPattern,
+              work.tableNameOrPattern);
+      validTxnList = getTxnMgr().getValidTxns();
+      long timeoutInMs = HiveConf.getTimeVar(conf,
+              HiveConf.ConfVars.REPL_BOOTSTRAP_DUMP_OPEN_TXN_TIMEOUT, TimeUnit.MILLISECONDS);
+      waitUntilTime = System.currentTimeMillis() + timeoutInMs;
+    }
 
     // TODO : instead of simply restricting by message format, we should eventually
     // move to a jdbc-driver-stype registering of message format, and picking message
     // factory per event to decode. For now, however, since all messages have the
     // same factory, restricting by message format is effectively a guard against
     // older leftover data that would cause us problems.
 
-    work.overrideEventTo(hiveDb);
+    work.overrideEventTo(hiveDb, bootDumpBeginReplId);
```

Comment:
work.maxEventLimit() should limit the events less than bootDumpBeginReplId. Need to set it properly.

Reply:
Nice catch. Thanks for pointing that out. Fixed.